Add design and registration

alexschneider opened the pull request:
User registration isn't completely finished yet, but the design and the application are functional, so this is probably OK to be merged if others are happy with it. 

Notable changes include:
- More ES6 goodness (with babel added back in to support it)
- Material Design!
- If a password doesn't match up, the username is preserved (this took far far too long to figure out how to do)
- Duplicated client side and server side validation

alexschneider committed:
Add design and registration

alexschneider commented:
BTW - to run the server you use 

```
npm start
```

alexschneider committed:
Fix tests

alexschneider committed:
Add working tests

mcoccia commented:
Why a 12 character min password? Seems high for me.
Also should it just be working when I do npm start? I type in my info and it just loads forever. I'm guessing it's just not hitting a server which is what you mean by it's not done yet. Also login gives me a 404 when I click on that tab which idk if we want things 404ing on master? Maybe we do idk.

alexschneider commented:
Yea, this probably shouldn't be merged quite yet. I'll get login and registration fully working later. I just wanted to see if everyone liked the design before committing to much more.

alexschneider committed:
Add more unit tests

alexschneider committed:
Add configuration and ORM

alexschneider committed:
Finally sequlize works (also logging!)

alexschneider committed:
Finish setting up user module and refactoring

alexschneider committed:
Set up registration, logging, and database

alexschneider committed:
Add secrets and finish registration

alexschneider commented:
This is probably ready to be merged. Please review!

alexschneider committed:
Remove useless jshintrc file

alexschneider commented on Leaderboard/bin/www:
```diff
@@ -3,88 +3,94 @@
-/**
- * Get port from environment and store in Express.
- */
+db().then(function() {
```
The database needs to be set up before certain parts of the application are set up. Since initializing the database doesn't block (seriously node?), we have to do the database setup prior to setting up the entire application, otherwise we run the risk of a potentially vulnerable race condition.

alexschneider commented on Leaderboard/helper.js:
```diff
@@ -0,0 +1,8 @@
+ * We need this so that express handles errors properly in async functions
+ * because express is dumb.
+ */
+export const asyncWrap = fn => (...args) => fn(...args).catch(args[2]);
```
I hate that we have to use this function, but express doesn't handle errors in asynchronous handlers properly (wtf express/node?)

alexschneider commented on Leaderboard/locals.js:
```diff
@@ -0,0 +1,16 @@
+const locals = {
+  // App configuration
+  appName: 'Fight or Flight',
+  primaryColor: 'indigo',
```
This should probably be moved to the config folder (default.yml), but these are all variables that are available to the templating engine. 

We can change the colors pretty easily by going [here](http://www.getmdl.io/customize/index.html) and picking the colors, and copying over to this file (someone more artistically minded than I should do this).

alexschneider commented on Leaderboard/logging.js:
```diff
@@ -0,0 +1,36 @@
+import expressWinston from 'express-winston';
+
+const winstonConfig = config.get('Logger');
+const loggerConfig = {
```
Most of these should probably be in the config file, but I was running into errors when I was doing that. Issue for later.

alexschneider commented on Leaderboard/package.json:
```diff
@@ -3,21 +3,32 @@
   "devDependencies": {
     "chai": "^3.2.0",
     "mocha": "^2.3.2",
+    "sqlite3": "^3.1.0",
```
In production, we will be using postgres, so this is why it's under dev dependencies.

alexschneider commented on Leaderboard/routes/users.js:
```diff
@@ -1,9 +1,100 @@
+  isEmailError: email,
+});
+
+const calculateSaltHash = password => {
```
Current calculations take approx 1.6 seconds. This is tunable in the configuration, but we're basically using a 64 byte salt, 512 byte hash, and 100k iterations. This is effectively unbreakable before the heat death of the universe. We can probably get it down to 100-200ms without compromising TOO much security, but I do like being overkill :D - opinions?

akrs replied:
Seems excessively high, could lead to DoS because our server is too busy with calculating hashes to actually serve requests.

alexschneider commented on Leaderboard/secret.js:
```diff
@@ -0,0 +1,16 @@
+import fs from 'fs';
+import config from 'config';
+import crypto from 'crypto';
+// TODO: Is this the best way to handle application secrets?
```
When an application first starts up, it will securely generate a 512 byte secret and store it in a file. I'm open to doing it other ways, but this way we don't have to have any user interaction when setting up the application, and it always behaves in a secure manner. Error handling probably isn't optimal, but I believe that if there is an error, it won't allow the application to start up. 

Also TODO: fix secret file permissions

alexschneider commented on Leaderboard/test/test_webserver.js:
```diff
@@ -1,16 +1,78 @@
-const app = require('../app'),
```
This file needs to be expanded. TODO

alexschneider commented on Leaderboard/views/index.hbs:
```diff
@@ -1,2 +1 @@
-<h1>{{title}}</h1>
-<p>Welcome to {{title}}</p>
+
```
We probably want to have some form of welcome page here explaining the application.

alexschneider commented:
It's not completely finished yet, but it's in a working state. I'd like to get it merged in so @jbarbosa1 can get email verification working. Once it gets merged in, I'll create issues for everything I mentioned and we'll go from there.

akrs commented:
Other than the password hashing issue, lgtm. :+1:

alexschneider commented:
What length of time spent hashing do you think would be acceptable?

akrs commented:
I'm thinking more like 100-150ms

alexschneider committed:
Reduce default number of iterations for encryption

Randomize the number of iterations to make it more difficult for
rainbow tables to be generated.

alexschneider commented:
Ok, I got it down to 250ms. I'm not comfortable with lowering it any further.

mcoccia commented:
Feels instant enough for me on the time. I think how it is now works

akrs commented:
:beers:
